File: openvz_exporter.rb
#!/opt/rbenv/shims/ruby

require 'sinatra'
require 'json'

set :bind, '::0'
set :port, 9119

class OpenVZPrometheus
  def vzlist
    JSON.parse(`vzlist -a -o hostname,ip,laverage -j`)
  end

  def to_prometheus(hostname, ip, value)
    "node_openvz_laverage{hostname=\"#{hostname}\", ip=\"#{ip}\"} #{value}\n"
  end

  def metrics
    vzlist.map do |vz|
      to_prometheus(vz['hostname'], vz['ip'].first, vz['laverage'].first)
    end
  end
end

get '/metrics' do
  content_type 'text/plain'
  OpenVZPrometheus.new.metrics
end

File: openvz_exporter.py
#!/usr/bin/python

import subprocess
import json
from flask import Flask
app = Flask(__name__)

def vzlist():
    command = "vzlist -o hostname,ip,laverage -j"
    p = subprocess.Popen(command, stdout=subprocess.PIPE, shell=True)
    return p.communicate()[0]

def to_prometheus(hostname, ip, value):
    return 'node_openvz_laverage{hostname="%s", ip="%s"} %s\n' % (hostname, ip, value)

@app.route("/metrics")
def metrics():
    output = []
    for vz in json.loads(vzlist()):
        output.append(to_prometheus(vz['hostname'], vz['ip'][0], vz['laverage'][0]))

    return "".join(output), 200, {'Content-type': 'text/plain'}

if __name__ == "__main__":
    app.run(host='::', port=9119)
